Issue 1635: Decypting PDF ebboks fixed

frickeo opened the pull request:
The decryption of pdf ebooks was broken (see https://github.com/apprenticeharper/DeDRM_tools/issues/1635).

If fixed this with two modifications

First of all, I got an empty list of objids when parsing a pdf document: In Python 2, the build-in function zip() returns a list and not an iterator. This changed in Python 3. I converted the iterator into a list again and got objids back.

Another correction was needed. to make the predictor code work again.

Commit:
Added a code to remove Kindle watermark.
https://github.com/matimatik/DeDRM_tools/commit/f3fbc3573eadc4b505ed296176f35cd5a917cdbf

tracyde committed:
Support Python 2.7 and Python 3 winreg imports on Windows

journeyman88 committed:
Update obok.py

Changed MAC address fetching code to address possibile regression

Commit:
Issue 1635: Decypting PDF ebboks fixed

Commit:
Issue 1635: Decypting PDF ebboks fixed

belm0 committed:
FAQ: note that Kindle 1.17 on Mac is 32-bit

ableeker committed:
Fix for decryption check

abort-retry-fail-ignore reviewed (commented):
Works for me. Tried with two problem pdfs, decrypted successfully. Thanks for tracking this down!

abort-retry-fail-ignore commented:
Thank you for this patch. Tested successfully on two pdfs.

gschenck commented:
Works. Thank you!

Leseratte10 commented:
I've found a PDF where this bugfix wasn't enough to make the DeDRM work again. 

There's code in the `nunpack` function that looks like this:
```
return struct.unpack('>L', b'\x00'+s)[0]
```

and that gave me some Python error about being unable to combine str and bytes. 
Changing that line to the following fixes this issue: 
```
return struct.unpack('>L', bytes([0]) + s)[0]
```

Maybe that can be added to this PR as well.

noDRM committed:
More PDF fixes

noDRM committed:
Remove broken CI

noDRM committed:
Add UUID to adobekey DER file names

noDRM committed:
Make keys fit into listbox

noDRM committed:
Update Readme

noDRM committed:
Improve key detection

noDRM committed:
Improve key detection for PDFs, too

noDRM committed:
Remove library book block

noDRM committed:
Remove library flag from MOBI book

noDRM committed:
Add old B&N algorihm (optional) just in case it's needed

noDRM committed:
Enable issue forms

noDRM committed:
Add IETF and Adobe font deobfuscation code

noDRM committed:
No longer break obfuscated fonts on DRM removal

noDRM committed:
Add useful error message for the new, uncracked ADEPT DRM

noDRM committed:
Add back Python2 support (ADEPT)

noDRM committed:
Fix ADE key import through plugin settings

noDRM committed:
Add B&N PDF DeDRM (untested), match UUID for Adobe PDFs

noDRM committed:
Remove CDP watermark from EPUBs

noDRM committed:
Auto-import keys from DeACSM plugin

noDRM committed:
Fix font deobfuscation for Python 2

noDRM committed:
Obok plugin cleanup

noDRM committed:
Add some more watermark removal code

noDRM committed:
Make CI auto-package the plugin

noDRM committed:
LCP support

noDRM committed:
Fixes a bug that sometimes caused the plugin to fail

Commit:
Fix typo in kindlekey.py that broke Mac version

Leseratte10 committed:
Fix username code for ADE key retrieval

noDRM committed:
Allow packaging without version number

noDRM committed:
Fix watermark stuff

noDRM committed:
Update to v10.0.2

noDRM committed:
Cleanup

noDRM committed:
Fix username decryption with unicode chars in Python2

noDRM committed:
Support importing multiple keys from ADE

noDRM committed:
Detect Kobo & Apple DRM in epubtest.py

noDRM committed:
Update changelog

noDRM committed:
Lots of B&N updates

noDRM committed:
Update PassHash documentation

noDRM committed:
Support for extracting PassHashes from ADE

noDRM committed:
Key retrieval updates

noDRM committed:
Edit .gitignore

noDRM committed:
Remove ancient code to import keys from ancient plugins

There were a couple specific DRM removal plugins before the DeDRM plugin
was created. These are obsolete since a long time, there's no need to
still have the code to import their config.

If people are still using these ancient plugins, they'll have to update
to an older version of DeDRM first, and then update to the current one.

noDRM committed:
Update watermark code

noDRM committed:
Ton of PDF DeDRM updates

- Support "Standard" and "Adobe.APS" encryptions
- Support decrypting with owner password instead of user password
- New function to return encryption filter name
- Support for V=5, R=5 and R=6 PDF files
- Support for AES256-encrypted PDF files
- Disable broken cross-reference streams in output

noDRM committed:
Fix broken key management

noDRM committed:
Update Changelog

noDRM committed:
Correct user pass padding, fix PDFStream export

ableeker committed:
Python 3 fixes

noDRM committed:
Cleanup

noDRM committed:
Begin work on standalone version

Now the plugin ZIP file (DeDRM_plugin.zip) can be run with a normal
Python interpreter as if it were a Python file (try
`python3 DeDRM_plugin.zip --help`). This way I can begin building a
standalone version (that can run without Calibre) without having to
duplicate a ton of code.

noDRM committed:
Bugfixes in standalone code for Calibre < 5 / Python 2

noDRM committed:
Add passhash interface to CLI

noDRM committed:
More work on standalone version, fix plugin

noDRM committed:
Try to fix B&N issues

noDRM committed:
Fix libcrypto DLL path search (see #13 and #14)

Co-authored-by: Adriano Caloiaro <code@adriano.fyi>

noDRM committed:
Fix FileNotFoundError during PassHash handling

noDRM committed:
Add "MemoryError" to FAQ

noDRM committed:
Make plugin work in Calibre 6 (Qt 6)

noDRM committed:
Fix IndexError in mobidedrm.py

noDRM committed:
Add empty placeholder file for LCP

noDRM committed:
Remove LCP references from Readme

noDRM committed:
Update FAQ

baby-bell committed:
Obok: Fix invalid UTF-8 causing UI to not open

For some reason, the title of a book on my device causes Obok to choke. Apparently it's not valid UTF-8.
This fixes that by ignoring decode errors.

a980e066a01 committed:
Remove OpenSSL support; only support PyCryptodome

This allows us to clean up the code a lot.

On Windows, it isn't installed by default and
most of the time not be found at all.

On M1 Macs, the kernel will kill the process instead.

Closes #33.

a980e066a01 committed:
Add support for "hardened" Adobe DRM

What took the most time was not reverse-engineering
the scheme, but actually finding books using it...

Closes #20, #25, #45

tkauppin reviewed (approved):
works fine for me

noDRM committed:
Update FAQs

Co-authored-by: ZolaLa <49111160+ZolaLa9@users.noreply.github.com>

noDRM committed:
Try to fix V3 PDF files

noDRM committed:
Debugging for __version issue

noDRM committed:
Make B&N plugin skip invalid hashes in Windows app

noDRM committed:
Improve error message

noDRM committed:
Update plugin readme

noDRM committed:
Add back unpad to fix Python2 support

noDRM committed:
(Hopefully) fix WineGetKeys for Kindle

noDRM committed:
Some untested Python2 Kindle bugfixes

noDRM committed:
Bugfix for Nook PDFs?

noDRM committed:
Fix print-replica Amazon books

noDRM committed:
Hopefully fix Kindle books

noDRM committed:
Fix Kindle for real

noDRM committed:
Fix Nook study key retrieval

sweetStreet committed:
fix the regular expression

noDRM committed:
Fix Obok plugin on Calibre 6 (#98)

noDRM committed:
Prepare release v10.0.3

noDRM committed:
Add some Python2 compat code I forgot to add earlier

noDRM committed:
Fix RSA.import_key (fixes #101)

Apparently "import_key" only exists in newer versions (as an alias to
"importKey"). "importKey" works in all versions ...

noDRM committed:
Fix Calibre 6 issue in Obok plugin

noDRM committed:
Fix plugin crash with invalid ADE key

noDRM committed:
Fix ZIP attribute "external_attr" getting reset

noDRM committed:
Fix PDF corruption in Calibre 4 (#104)

noDRM committed:
Delete unused files

noDRM committed:
Another Python2 Bugfix for Obok

noDRM committed:
Move SafeUnbuffered to own Python file

noDRM committed:
Couple Python 2 fixes in (unsupported) standalone scripts

noDRM committed:
Remove AlfCrypto libraries and perform everything in Python

The old AlfCrypto DLL, SO and DYLIB files are ancient,
I don't have the systems to recompile them all, they
cause issues on ARM Macs, and I doubt with all the Python
improvements over the last years that they have a significant
performance advantage. And even if that's the case, nobody is
importing hundreds of DRM books at the same time so it shouldn't
hurt if some decryptions might take a bit longer.

noDRM committed:
Move unicode_argv to its own file

noDRM committed:
Cleanup

noDRM committed:
More stuff I missed

noDRM committed:
Cleanup / SafeUnbuffered bugfix

noDRM committed:
Fix DeACSM import for PDF files

noDRM committed:
Fix tons of PDF-related issues

RolandWH committed:
fix 2 spelling errors in FAQs.md

noDRM committed:
Fix key import sometimes generating corrupted keys.

Should fix #145, #134, #119, #116, #115, #109 and maybe others.

noDRM committed:
Cleanup

noDRM committed:
Update README, fixes #136

noDRM committed:
Update changelog

noDRM committed:
Add support for empty arrays (<>) in PDF objects. Fixes #183.

noDRM committed:
Strip whitespace from Kindle serials (#158)

noDRM committed:
Untested code for the Obok plugin to allow adding duplicate books.

See #148

Commit:
Merge remote-tracking branch 'upstream/master' into master

noDRM committed:
Fix exception in error logging in ineptpdf

noDRM committed:
Fix Python3 bug in stylexml2css.py, fixes #232

noDRM committed:
PDF: Ignore invalid objid in non-strict mode, fixes #233

noDRM committed:
Update CHANGELOG

noDRM committed:
Try fixing a Python2 bug in the Obok plugin (#235)

noDRM committed:
Fix issue with remaining data in encryption.xml

noDRM committed:
Support for Adobe's 'aes128-cbc-uncompressed' encryption (see #242)

noDRM committed:
Try to add support for new K4PC

Co-authored-by: Andrew Innes <andrew.c12@gmail.com>
Co-authored-by: Satsuoni <satsuoni@hotmail.com>

noDRM committed:
Fix font decryption with multiple IDs (#347)

noDRM committed:
Fix python2 issues in Kindle and Nook code (#355)

noDRM committed:
Added fix for padding being correct on accident

Co-authored-by: Satsuoni <satsuoni@hotmail.com>

Satsuoni committed:
Added several more scramble functions to Kindle decrypt

noDRM committed:
Update CI to create an automatic beta release

noDRM committed:
Disable auto-prerelease again (#358)

Commit:
Merge remote-tracking branch 'upstream/master'
